subprojects/core/src/main/java/org/gradle/api/attributes/Attribute.java

```diff
@@ -117,4 +119,30 @@ public int hashCode() {
     public String toString() {
         return name;
     }
+
+    private static <T> void validateType(String name, Class<T> clazz) {
+        if (clazz.isArray()) {
+            validateType(name, clazz.getComponentType());
+            return;
+        }
+        if (Scalars.isScalarType(clazz)) {
+            return;
+        }
+        if (Named.class.isAssignableFrom(clazz)) {
```

Comment:
I don't think we isolate attribute values anywhere yet. It wouldn't be necessary if they are effectively immutable, but there are 2 cases where they would:

- the type of the attribute is an array
- the type of the attribute implements `Named`, and the user brought a value _without_ constructing it using `objectFactory`